src/Http/Routing/src/RouteEndpointDataSource.cs

```diff
@@ -0,0 +1,262 @@
+        {
+            endpoints.Add((RouteEndpoint)CreateRouteEndpointBuilder(entry, context.Prefix, context.Conventions).Build());
+        }
+        return endpoints;
+    }
+
+    public override IChangeToken GetChangeToken() => NullChangeToken.Singleton;
+
+    // For testing
+    internal RouteEndpointBuilder GetSingleRouteEndpointBuilder()
+    {
+        if (_routeEntries.Count is not 1)
+        {
+            throw new InvalidOperationException($"There are {_routeEntries.Count} endpoints defined! This can only be called for a single endpoint.");
+        }
+
+        return CreateRouteEndpointBuilder(_routeEntries[0]);
+    }
+
+    [UnconditionalSuppressMessage("Trimmer", "IL2026",
+        Justification = "We surface a RequireUnreferencedCode in the call to Map method adding this EndpointDataSource. " +
+                        "The trimmer is unable to infer this.")]
+    private RouteEndpointBuilder CreateRouteEndpointBuilder(RouteEntry entry, RoutePattern? groupPrefix = null, IReadOnlyList<Action<EndpointBuilder>>? groupConventions = null)
+    {
+        var pattern = RoutePatternFactory.Combine(groupPrefix, entry.RoutePattern);
+        var handler = entry.RouteHandler;
+        var displayName = pattern.RawText ?? pattern.DebuggerToString();
+
+        // Methods defined in a top-level program are generated as statics so the delegate target will be null.
+        // Inline lambdas are compiler generated method so they be filtered that way.
+        if (GeneratedNameParser.TryParseLocalFunctionName(handler.Method.Name, out var endpointName)
+            || !TypeHelper.IsCompilerGeneratedMethod(handler.Method))
+        {
+            endpointName ??= handler.Method.Name;
+            displayName = $"{displayName} => {endpointName}";
+        }
+
+        if (entry.HttpMethods is not null)
+        {
+            // Prepends the HTTP method to the DisplayName produced with pattern + method name
+            displayName = $"HTTP: {string.Join(", ", entry.HttpMethods)} {displayName}";
+        }
+
+        if (entry.IsFallback)
+        {
+            displayName = $"Fallback {displayName}";
+        }
+
+        RequestDelegate? factoryCreatedRequestDelegate = null;
+        RequestDelegate redirectedRequestDelegate = context =>
+        {
+            if (factoryCreatedRequestDelegate is null)
+            {
+                throw new InvalidOperationException(Resources.RouteEndpointDataSource_RequestDelegateCannotBeCalledBeforeBuild);
+            }
+
+            return factoryCreatedRequestDelegate(context);
+        };
+
+        // The Map methods don't support customizing the order apart from using int.MaxValue to give MapFallback the lowest priority.
+        // Otherwise, we always use the default of 0 unless a convention changes it later.
+        var order = entry.IsFallback ? int.MaxValue : 0;
+
+        RouteEndpointBuilder builder = new(redirectedRequestDelegate, pattern, order)
```

Comment:
So thinking about this more I think I'd like to see one change to clean up how we're handling filters currently. Instead of treating it like metadata that we have to hunt from the collection, create a `RouteHandlerEndpointBuilder` that has the list of filters instead (we would unseal RouteEndpointBuilder).

```C#
public sealed class RouteHandlerEndpointBuilder : RouteEndpointBuilder
{
         public IReadOnlyList<Func<RouteHandlerContext, RouteHandlerFilterDelegate, RouteHandlerFilterDelegate>>? RouteHandlerFilterFactories { get; init; }
  
     public RouteHandlerEndpointBuilder(
       RequestDelegate requestDelegate,
       RoutePattern routePattern,
       int order) : base(requestDelegate, routePattern, order) { }
}
```

I like this for 2 reasons:
- It's a pattern for identifying these endpoints specifically so we can choose to skip applying metadata to endpoints based on this type.
- No weird merging of metadata entries is required.

Reply:
That's a really good idea. We'd have to unseal `RouteEndpointBuilder`, but big deal. I think we can do one better though.

We could consider adding `RouteHandlerFilterFactories` directly to `RouteEndpointBuilder`. 

`RequestDelegate` is just a special case of our `Delegate` handling from a functional perspective. We should still keep things as they are today if there are no filters to avoid any possible startup penalty and trimming issues. But if someone does apply a filter to a `RequestDelegate` call to `MapPost(...)`, etc..., they should get to have their filter run on it if they want to.

I know this also means if someone calls something `MapControllers()`, `MapHub<Hub>` in a filtered group, the filters will run on those endpoints too, but I think most developers will actually really enjoy seeing the all registered routes and stuff in there filters if they filter an entire group. It's easy to separate Controllers and Hubs from other routes in the group so they don't get filtered. It's also easy enough for a filter to skip over route handlers `Delegate`s that are really just `RequestDelegate`s if they want to and that way they incur no per-request performance penalty.

This would involve calling RDF.Create() in `RouteEndpointBuilder.Build()` if and only if there are filters set. I know it might sound like going to far, but I really think it's super powerful and worth doing if we make it clear this is how filters work, and make it very easy to choose to no-op in your filter for plain `RequestDelegate`'s if you want to and incur no per request performance penalty. People writing their own debugging and diagnostic tools are going to love it. I might write a few now to show it off.

Reply:
I've done this. See [this commit](https://github.com/dotnet/aspnetcore/pull/42195/commits/1aa70bb60e8d64b748c8fb8a45cbbf121b3ce2fb) for most of the implementation. Absolutely zero cost per-request unless a filter is added directly to the `RouteEndpointBuilder` *and* the filter actually modifies the per-invocation `RouteHandlerFilterDelegate`. If the filter does modify `RouteHandlerFilterDelegate` for the `RequestDelegate` endpoint, the modified `RouteHandlerFilterDelegate` does run of course, but that's what you asked for. A `RequestDelegate` is just one kind a `Delegate` after all. And aside from applying filters and handling `Task<T>`-returning `RequestDelegate`s (for more than just `MapGet` after that commit I might add), RDF does not change the behavior of the `RequestDelegate` compared to just running it directly.

This is all new API in .NET 7. No one is forcing anyone to `AddRouteHandlerFilter`s to everyday plain-`RequestDelegate` backed `RouteEndpointBuilder`s, but if you do, it's' very nice to have them run. What's really nice is this allows filters to run on practically any endpoint in existence. `MapHub`, `MapHealthChecks`, `MapControllers`, `MapRazorPages`, etc... But again, only if you ask for it. This is the first time people can even try doing this now that `AddRouteHandlerFilter` can be applied to any `IEndpointConventionBuilder`.

Reply:
I think this should be decoupled from this PR and the implementation shouldn't use RDF.

Reply:
Throwing in my two cents here now that I've reviewed the change. While I can recognize that there might be compelling scenarios for it, I'm a little cautious about:

- Changing our direction on whether or not route handler-filters could be applied to non-route handlers. We were committed enough to that stance to rename the API from `AddFilter` to `AddRouteHandlerFilter`. Side note: if we do want to pursue this direction then we should consider changing it back to `AddFilter`. 
- Invoking the RDF in all route endpoint builders. This change accounts invoking RDF from non-route handler scenarios when there are filters involved but have we considered other changes in the RDF that need to be made so that it can be invoked in more places.

In general, this seems high-risk/medium-reward and would be more sensible to approach as an independent change.